daemon/regmaild/regmaild.c

```diff
@@ -516,87 +704,204 @@ regcheck_ambiguous_id(const char *userid)
+{
+    if (req->cb != sizeof(*req)) {
+        fprintf(stderr, "error: corrupted request, cb: %lu != %lu\n",
+                req->cb, sizeof(*req));
+        return false;
+    }
+
+    switch (req->operation)
+    {
+        case REGMAILDB_REQ_COUNT:
+        case REGMAILDB_REQ_SET:
+            if (!*req->userid || !*req->email) {
+                fprintf(stderr, "invalid request: op=[%d], uid=[%s]\n",
+                        req->operation,
+                        req->userid);
+                return false;
+            }
+            break;
+
+        case REGCHECK_REQ_AMBIGUOUS:
+            // No checks.
+            break;
+
+        default:
+            fprintf(stderr, "unknown request: op=[%d]\n", req->operation);
+            return false;
+    }
+    return true;
+}
+
+static bool
+validate_verifydb_request(const verifydb_req *req)
+{
+    if (req->cb != sizeof(*req)) {
+        fprintf(stderr, "error: corrupted request, cb: %lu != %lu\n",
+                req->cb, sizeof(*req));
+        return false;
+    }
+    if (!check_null_terminated(req->userid, sizeof(req->userid))) {
+        fprintf(stderr, "invalid request: op=[%d], userid is not null-terminated\n",
+                req->operation);
+        return false;
+    }
+    if (!*req->userid) {
+        fprintf(stderr, "invalid request: op=[%d], empty userid\n",
+                req->operation);
+        return false;
+    }
+    if (req->operation == VERIFYDB_REQ_COUNT ||
+        req->operation == VERIFYDB_REQ_SET) {
+        if (req->vmethod == VMETHOD_UNSET) {
+            fprintf(stderr, "invalid request: op=[%d], vmethod is not set\n",
+                    req->operation);
+            return false;
+        }
+        if (!check_null_terminated(req->vkey, sizeof(req->vkey))) {
+            fprintf(stderr, "invalid request: op=[%d], vkey is not null-terminated\n",
+                    req->operation);
+            return false;
+        }
+    }
+    return true;
 }
 
 static void 
 client_cb(int fd, short event, void *arg)
 {
     struct event *ev = (struct event*) arg;
-    regmaildb_req req = {0};
-    int ret = -1;
+    regmaildb_req_storage storage = {};
+    regmaildb_req_header *header = &storage.m.header;
+    int8_t *rest = (int8_t *)&storage;
+    rest += sizeof(*header);
 
     assert(ev);
 
-    if ( (event & EV_TIMEOUT) ||
-        !(event & EV_READ) ||
-         toread(fd, &req, sizeof(req)) != sizeof(req) ||
-         req.cb != sizeof(req))
+    if ((event & EV_TIMEOUT) ||
+            !(event & EV_READ) ||
+            toread(fd, header, sizeof(*header)) != sizeof(*header) ||
+            header->cb < sizeof(*header) ||
+            header->cb > sizeof(storage) ||
+            toread(fd, rest, header->cb - sizeof(*header)) != header->cb - sizeof(*header))
     {
-        fprintf(stderr, "error: corrupted request.\r\n");
-        close(fd);
-        free(ev);
-        return;
+        fprintf(stderr, "error: corrupted request.\n");
```

Comment:
Either leave the carriage return alone for now, or make a patch removing it from all log messages.
It is unrelated to what the commit log says the change is supposed to do.
(Why were there carriage returns in the first place...)

Reply:
Made a separate commit that removes "\r".